src/jit/gentree.cpp

```diff
@@ -6814,105 +7071,112 @@ void Compiler::gtBlockOpInit(
+        {
+            // Make this a NOP
+            // TODO-Cleanup: probably doesn't matter, but could do this earlier and avoid creating a GT_ASG
+            result->gtBashToNOP();
+            return;
+        }
+    }
 
     // Propagate all effect flags from children
-    result->gtFlags |= result->gtOp.gtOp1->gtFlags & GTF_ALL_EFFECT;
+    result->gtFlags |= dst->gtFlags & GTF_ALL_EFFECT;
     result->gtFlags |= result->gtOp.gtOp2->gtFlags & GTF_ALL_EFFECT;
 
+    // TODO-1stClassStructs: This should be done only if the destination is non-local.
     result->gtFlags |= (GTF_GLOB_REF | GTF_ASG);
 
     // REVERSE_OPS is necessary because the use must occur before the def
-    result->gtOp.gtOp1->gtFlags |= GTF_REVERSE_OPS;
-
-    if (result->gtOper == GT_INITBLK)
-    {
-        result->gtFlags |= (dst->gtFlags & GTF_EXCEPT) | (hndOrSize->gtFlags & GTF_EXCEPT);
-    }
-    else
-    {
-        result->gtFlags |=
-            (dst->gtFlags & GTF_EXCEPT) | (srcOrFillVal->gtFlags & GTF_EXCEPT) | (hndOrSize->gtFlags & GTF_EXCEPT);
+    result->gtFlags |= GTF_REVERSE_OPS;
 
-        // If the class being copied contains any GC pointer we store a class handle
-        // and we must set the flag GTF_BLK_HASGCPTR, so that the register predictor
-        // knows that this GT_COPYBLK will use calls to the ByRef Assign helper
-        //
-        if ((hndOrSize->OperGet() == GT_CNS_INT) && hndOrSize->IsIconHandle(GTF_ICON_CLASS_HDL))
-        {
-            hndOrSize->gtFlags |= GTF_DONT_CSE; // We can't CSE the class handle
-            result->gtFlags |= GTF_BLK_HASGCPTR;
-        }
-    }
+    result->gtFlags |= (dst->gtFlags & GTF_EXCEPT) | (srcOrFillVal->gtFlags & GTF_EXCEPT);
 
-    if (volatil)
+    if (isVolatile)
     {
         result->gtFlags |= GTF_BLK_VOLATILE;
     }
 
 #ifdef FEATURE_SIMD
-    if (oper == GT_COPYBLK && srcOrFillVal->OperGet() == GT_ADDR && dst->OperGet() == GT_ADDR)
+    if (result->OperIsCopyBlkOp() && varTypeIsSIMD(srcOrFillVal))
     {
         // If the source is a GT_SIMD node of SIMD type, then the dst lclvar struct
         // should be labeled as simd intrinsic related struct.
         // This is done so that the morpher can transform any field accesses into
         // intrinsics, thus avoiding conflicting access methods (fields vs. whole-register).
 
-        GenTreePtr srcChild = srcOrFillVal->gtGetOp1();
-        GenTreePtr dstChild = dst->gtGetOp1();
-
-        if (dstChild->OperIsLocal() && varTypeIsStruct(dstChild) && srcChild->OperGet() == GT_SIMD &&
-            varTypeIsSIMD(srcChild))
+        GenTree* src = srcOrFillVal;
+        if (src->OperIsIndir() && (src->AsIndir()->Addr()->OperGet() == GT_ADDR))
         {
-            unsigned   lclNum                = dst->gtGetOp1()->AsLclVarCommon()->GetLclNum();
-            LclVarDsc* lclVarDsc             = &lvaTable[lclNum];
-            lclVarDsc->lvUsedInSIMDIntrinsic = true;
+            src = src->AsIndir()->Addr()->gtGetOp1();
+        }
```

Comment:
On looking at this further, the logic here is simply getting the source for setting the `lvUsedInSIMDIntrinsic` flag, whereas the logic in `gtNewBlkOpNode` is modifying what will be used as the source in the assignment. While there is probably an opportunity for simplification, I'm not inclined to add the SIMD check to `impAssignStructPtr`, which is already rather complex, so I think it's best to leave this here.

---
In reply to: [77283021](https://github.com/dotnet/coreclr/pull/7027#discussion_r77283021) [](ancestors = 77283021,77280046)